packages/react-query-next-experimental/src/ReactQueryStreamedHydration.tsx

```diff
@@ -0,0 +1,89 @@
+'use client'
+
+import type { ContextOptions, DehydratedState } from '@tanstack/react-query'
+import { dehydrate, hydrate, useQueryClient } from '@tanstack/react-query'
+import * as React from 'react'
+import type { HydrationStreamProviderProps } from './HydrationStreamProvider'
+import { createHydrationStreamProvider } from './HydrationStreamProvider'
+
+const stream = createHydrationStreamProvider<DehydratedState>()
+
+/**
+ * This component is responsible for:
+ * - hydrating the query client on the server
+ * - dehydrating the query client on the server
+ */
+export function ReactQueryStreamedHydration(props: {
+  children: React.ReactNode
+  context?: ContextOptions['context']
+  transformer?: HydrationStreamProviderProps<DehydratedState>['transformer']
+}) {
+  const queryClient = useQueryClient({
+    context: props.context,
+  })
+
+  // <server only>
+  const isSubscribed = React.useRef(false)
+  /**
+   * We need to track which queries were added/updated during the render
+   */
+  const [trackedKeys] = React.useState(() => new Set<string>())
+  /**
+   * Track which queries were already passed to the client so we don't pass them again
+   */
+  // const [passedKeys] = useState(() => new Set<string>());
+
+  const cache = queryClient.getQueryCache()
+
+  if (typeof window === 'undefined' && !isSubscribed.current) {
+    // Do we need to care about unsubscribing? I don't think so to be honest
+    cache.subscribe((event) => {
```

Comment:
I thiiink this is fine as its only mounted once on the server and the GC should clean it up once it is rendered, but might be worth looking into? There is no cleanup callback thingie on the server afaik

Reply:
I think it's only troublesome if it cannot be GC'd for some reason, e.g. a running timer. `cache.subscribe` per se doesn't create those, so I also think it's fine.

Reply:
Side effect in render which directly mutates state and also doesn't unsubscribe? Hat-trick, I like it. 😉 

I don't see `isSubscribed.current` ever being set to `true` though, forgot?

Reply:
you're right - we're never setting the `isSubscribed` ref. This bears the question: do we even need it? Why would that render function run more than once on the server?

Reply:
If it renders inside a Suspense boundary that suspends I'm not sure what happens or if behaviour is guaranteed, so it does feel safer and with no downsides to use it I think? 😄

Reply:
hm, here's my thinking:

1) the way I understand suspense is that it unmounts the component, then re-mounts it. At that point, the state of the ref would be lost anyways.
2) The `ReactQueryStreamedHydration` likely goes at the very top of your component tree, like in the example:
https://github.com/TanStack/query/blob/6ebe4f347a05462ef8c32d360fa183a4079b7515/examples/react/nextjs-suspense-streaming/src/app/providers.tsx#L22-L27
3) Even if we were to double subscribe - it would just add the same hash to the `trackedKeys` Set, which doesn't do anything.

> with no downsides to use it I think?

The downside is that we have some code that we have to maintain and ship to the browser, even though we're not really sure if we need it 😅 .

I've removed it here: 4ef87c15

But if you think that we should re-add it, we can. But yes, in that case, we would need to set the value to `true` in that if branch.

Reply:
Hm, oh yeah, it does loose it's values on re-render. Fair enough!